src/sql/src/catalog.rs

```diff
@@ -368,13 +369,33 @@ impl fmt::Display for CatalogItemType {
 
 /// Details about a type in the catalog.
 #[derive(Clone, Debug, Eq, PartialEq)]
-pub struct CatalogTypeDetails {
-    //TODO(jkosh44) removing this would remove cyclic dependencies between types and make it
-    // easier to dynamically assign builtin type ids.
+pub struct CatalogTypeDetails<T: TypeReference> {
     /// The ID of the type with this type as the array element, if available.
     pub array_id: Option<GlobalId>,
     /// The description of this type.
-    pub typ: CatalogType,
+    pub typ: CatalogType<T>,
+}
+
+/// Represents a reference to type in the catalog
+pub trait TypeReference {
```

Comment:
Do you actually need a trait here? I would have expected `CatalogType<&'static string>` and `CatalogType<GlobalId>` to work on their own!

Reply:
Eh, nvm, it does seem to make the usage of the parameter much clearer!

Reply:
Yeah I don't think it's needed, I was just trying to make it explicit what I actually expected to be used as a reference. I somewhat modeled it after `AstInfo`, and I thought it may even be possible to re-use `AstInfo::DataType`, but I didn't want to have to mess with wrapping everything in an `Ident`. Also I figured this was non-invasive enough.